src/main/java/de/tum/in/www1/artemis/domain/participation/ProgrammingExerciseTeamParticipation.java

```diff
@@ -0,0 +1,105 @@
+package de.tum.in.www1.artemis.domain.participation;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+
+import javax.persistence.Column;
+import javax.persistence.DiscriminatorValue;
+import javax.persistence.Entity;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.annotation.JsonView;
+
+import de.tum.in.www1.artemis.domain.Exercise;
+import de.tum.in.www1.artemis.domain.ProgrammingExercise;
+import de.tum.in.www1.artemis.domain.view.QuizView;
+
+@Entity
+@DiscriminatorValue(value = "PETP")
+public class ProgrammingExerciseTeamParticipation extends TeamParticipation implements ProgrammingExerciseParticipation {
+
+    private static final long serialVersionUID = 1L;
+
+    @Column(name = "repository_url")
+    @JsonView(QuizView.Before.class)
+    private String repositoryUrl;
+
+    @Column(name = "build_plan_id")
+    @JsonView(QuizView.Before.class)
+    private String buildPlanId;
+
+    public String getRepositoryUrl() {
+        return repositoryUrl;
+    }
+
+    public Participation repositoryUrl(String repositoryUrl) {
+        this.repositoryUrl = repositoryUrl;
+        return this;
+    }
+
+    public void setRepositoryUrl(String repositoryUrl) {
+        this.repositoryUrl = repositoryUrl;
+    }
+
+    public String getBuildPlanId() {
+        return buildPlanId;
+    }
+
+    public Participation buildPlanId(String buildPlanId) {
+        this.buildPlanId = buildPlanId;
+        return this;
+    }
+
+    public void setBuildPlanId(String buildPlanId) {
+        this.buildPlanId = buildPlanId;
+    }
+
+    /**
+     * @return the repository URL as an URL Object
+     */
+    @JsonIgnore
+    public URL getRepositoryUrlAsUrl() {
```

Comment:
All the methods below copy the ones from `ProgrammingExerciseStudentParticipation`. Can you move them out to a separate class and use them in both classes?

Reply:
I think in this case we'll have to bite the bullet and live with the code duplication since multiple inheritance is not really supported in Java. If we were to extract out the shared code into a new abstract class `ProgrammingExerciseParticipantParticipation`, the class `ProgrammingExerciseStudentParticipation` would need to inherit from both the new class and from `StudentParticipation` (the case for `ProgrammingExerciseTeamParticipation` is analogous).

We actually already have this exact case in the current codebase: It is "solved" by having the interface `ProgrammingExerciseParticipation` that gets implemented by `ProgrammingExerciseStudentParticipation`, `SolutionProgrammingExerciseParticipation` and `TemplateProgrammingExerciseParticipation`. These three classes all pretty much duplicate the same code. Consistency is enforced via the interface but the code duplication can't really be avoided unfortunately :/

Reply:
Yes, **but** we could still solve this problem. Java has default interface methods, which were originally introduced in order to provide backwards compatibility when adding new methods to interfaces. Normally, one should use an abstract class in most cases for default implementations. However, since we already have this interesting "`ProgrammingExerciseParticipation` interface solution" here, you could move all of this code into default methods in the interface.

Reply:
By all of this code do you mean `getRepositoryUrlAsUrl`? I think most of the other code can't be moved since default methods in interfaces can't access instance variables.